[WebGPU] logic for primitive restarts is incorrectly applied to non-strip formats

mwyrzykowski opened the pull request:
#### f58a434b02ddcd5a3142f52ee3c6c2506e569fc9
<pre>
[WebGPU] logic for primitive restarts is incorrectly applied to non-strip formats
<a href="https://bugs.webkit.org/show_bug.cgi?id=274731">https://bugs.webkit.org/show_bug.cgi?id=274731</a>
&lt;radar://128769041&gt;

Reviewed by Tadeu Zagallo.

If a non-strip index buffer was used in a vertex buffer call with
2^16 - 1 vertices, we would incorrectly zero-out the draw call, thinking
0xffff was the primitive restart value.

Technically also applies to uint32_t index buffers, but we don&apos;t allow
creating a vertex buffer with 2^32 - 1 vertices so that correction is applied
in this patch here but not possible to encounter in practice.

* Source/WebGPU/WebGPU/Device.mm:
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::clampIndexBufferToValidValues):
(WebGPU::RenderPassEncoder::clampIndirectIndexBufferToValidValues):

* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::drawIndexedIndirect):

Canonical link: <a href="https://commits.webkit.org/279444@main">https://commits.webkit.org/279444@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/38bfdce7cb8e51ead0c2a300189580831f310f44

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32843 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5993 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56767 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4213 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40298 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4000 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43353 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2761 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55585 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31036 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46223 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24489 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27888 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2369 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49648 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3715 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58362 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28644 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3720 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50755 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29851 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/46420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50101 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11667 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30780 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29622 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -582,15 +582,17 @@ static void setOwnerWithIdentity(id<MTLResourceSPI> resource, auto webProcessID)
-        ushort plusOneValue = 1 + indexBuffer[indexId]; // plusOne to handle primitive restarts
-        if (indexedOutput.baseVertex >= vertexCount[0] || plusOneValue + indexedOutput.baseVertex >= vertexCount[0] + 1)
+    [[vertex]] void vsUshort(device const ushort* indexBuffer [[buffer(0)]], device MTLDrawIndexedPrimitivesIndirectArguments& indexedOutput [[buffer(1)]], const constant uint* data [[buffer(2)]], uint indexId [[vertex_id]]) {
+        ushort plusOneValue = data[primitiveRestart] + indexBuffer[indexId]; // plusOne to handle primitive restarts
```
It looks like the comment/variable name might no longer apply

tadeuzagallo commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -582,15 +582,17 @@ static void setOwnerWithIdentity(id<MTLResourceSPI> resource, auto webProcessID)
-        uint plusOneValue = 1 + indexBuffer[indexId]; // plusOne to handle primitive restarts
-        if (indexedOutput.baseVertex >= vertexCount[0] || plusOneValue + indexedOutput.baseVertex >= vertexCount[0] + 1)
+    [[vertex]] void vsUint(device const uint* indexBuffer [[buffer(0)]], device MTLDrawIndexedPrimitivesIndirectArguments& indexedOutput [[buffer(1)]], const constant uint* data [[buffer(2)]], uint indexId [[vertex_id]]) {
+        uint plusOneValue = data[primitiveRestart] + indexBuffer[indexId]; // plusOne to handle primitive restarts
```
ditto

mwyrzykowski replied:
oh indeed, I will correct this, thanks for catching

mwyrzykowski committed:
[WebGPU] logic for primitive restarts is incorrectly applied to non-strip formats
https://bugs.webkit.org/show_bug.cgi?id=274731
<radar://128769041>

Reviewed by Tadeu Zagallo.

If a non-strip index buffer was used in a vertex buffer call with
2^16 - 1 vertices, we would incorrectly zero-out the draw call, thinking
0xffff was the primitive restart value.

Technically also applies to uint32_t index buffers, but we don't allow
creating a vertex buffer with 2^32 - 1 vertices so that correction is applied
in this patch here but not possible to encounter in practice.

* Source/WebGPU/WebGPU/Device.mm:
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::clampIndexBufferToValidValues):
(WebGPU::RenderPassEncoder::clampIndirectIndexBufferToValidValues):

* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::drawIndexedIndirect):

Canonical link: https://commits.webkit.org/279444@main
